File: python/basics/2-pointer/125_valid_palindrome.py
#https://leetcode.com/problems/valid-palindrome/
'''
2-pointe technique
isalnum()
'''

def valid_palindrome(s):
    N=len(s)
    l = 0
    r = N-1
    while l < r:
        if not s[l].isalnum():
            l += 1
            continue
        if not s[r].isalnum():
            r -= 1
            continue

        if s[l].lower() != s[r].lower():  #you forgot to change the case !@#$
            return False
        l += 1
        r -= 1
    return True




s = "A man, a plan, a canal: Panama"
result = valid_palindrome(s)
print("result: " + str(result))
#Output: true

s = "race a car"
#Output: false
result = valid_palindrome(s)
print("result: " + str(result))

s = " "
#Output: true
result = valid_palindrome(s)
print("result: " + str(result))

File: problem_solving/2-pointer/cpp/125_valid_palindrome.cpp
/*
 
 * 
 */


#include<string>
#include<cctype>
#include<iostream>
using namespace std;

bool is_palindrome(const std::string& s) {
    int l = 0;
    int r = s.length()-1;
    while(l < r) {
        while(l < r && !std::isalnum(s[l])) {
            l += 1;
        }
        while(l < r && !std::isalnum(s[r])) {
            r -= 1;
        }
        if(l < r && std::tolower(s[l]) != std::tolower(s[r])) {
            return false;
        }
        l += 1;
        r -= 1;
    }
    return true;
}


int main() {
    std::string s = "A man, a plan, a canal: Panama";
    //Output: true
    std::cout << "is_palindrome:" << std::endl;
    bool result = is_palindrome(s);
    cout << result << std::endl;

    s = "race a car";
    //Output: false
    std::cout << "is_palindrome:" << std::endl;
    result = is_palindrome(s);
    cout << result << std::endl;


    s = " ";
    //Output: true
    std::cout << "is_palindrome:" << std::endl;
    result = is_palindrome(s);
    cout << result << std::endl;

    return 0;
}